packages/components/src/navigable-container/container.js

```diff
@@ -0,0 +1,132 @@
+	constructor() {
+		super( ...arguments );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.bindContainer = this.bindContainer.bind( this );
+
+		this.getFocusableContext = this.getFocusableContext.bind( this );
+		this.getFocusableIndex = this.getFocusableIndex.bind( this );
+	}
+
+	bindContainer( ref ) {
+		const { forwardedRef } = this.props;
+		this.container = ref;
+
+		if ( isFunction( forwardedRef ) ) {
+			forwardedRef( ref );
+		} else if ( forwardedRef && 'current' in forwardedRef ) {
+			forwardedRef.current = ref;
+		}
+	}
+
+	getFocusableContext( target ) {
+		const { onlyBrowserTabstops } = this.props;
+		const finder = onlyBrowserTabstops ? focus.tabbable : focus.focusable;
+		const focusables = finder.find( this.container );
+
+		const index = this.getFocusableIndex( focusables, target );
+		if ( index > -1 && target ) {
+			return { index, target, focusables };
+		}
+		return null;
+	}
+
+	getFocusableIndex( focusables, target ) {
+		const directIndex = focusables.indexOf( target );
+		if ( directIndex !== -1 ) {
+			return directIndex;
+		}
+	}
+
+	onKeyDown( event ) {
+		if ( this.props.onKeyDown ) {
+			this.props.onKeyDown( event );
+		}
+
+		const { getFocusableContext } = this;
+		const { cycle = true, eventToOffset, onNavigate = noop, stopNavigationEvents } = this.props;
+
+		const offset = eventToOffset( event );
+
+		// eventToOffset returns undefined if the event is not handled by the component
+		if ( offset !== undefined && stopNavigationEvents ) {
+			// Prevents arrow key handlers bound to the document directly interfering
+			event.nativeEvent.stopImmediatePropagation();
+
+			// When navigating a collection of items, prevent scroll containers
+			// from scrolling.
+			if ( event.target.getAttribute( 'role' ) === 'menuitem' ) {
+				event.preventDefault();
+			}
+
+			event.stopPropagation();
+		}
+
+		if ( ! offset ) {
+			return;
+		}
+
+		const context = getFocusableContext( document.activeElement );
+		if ( ! context ) {
+			return;
+		}
+
+		const { index, focusables } = context;
+		const nextIndex = cycle ? cycleValue( index, focusables.length, offset ) : index + offset;
+		if ( nextIndex >= 0 && nextIndex < focusables.length ) {
+			focusables[ nextIndex ].focus();
+			onNavigate( nextIndex, focusables[ nextIndex ] );
+		}
+	}
+
+	render() {
+		const { children, ...props } = this.props;
+		// Disable reason: Assumed role is applied by parent via props spread.
+		/* eslint-disable jsx-a11y/no-static-element-interactions */
+		return (
+			<div ref={ this.bindContainer }
+				{ ...omit( props, [
+					'stopNavigationEvents',
+					'eventToOffset',
+					'onNavigate',
+					'cycle',
+					'onlyBrowserTabstops',
+					'forwardedRef',
+				] ) }
+				onKeyDown={ this.onKeyDown }
+				onFocus={ this.onFocus }>
+				{ children }
+			</div>
+		);
+	}
+}
+
+const forwardedNavigableContainer = ( props, ref ) => {
+	return <NavigableContainer { ...props } forwardedRef={ ref } />;
```

Comment:
Minor: I think the `ownProps` pattern we've adopted elsewhere (e.g. [`withSelect`](https://github.com/WordPress/gutenberg/blob/14278906c763f14482ecb81555b7bde437f5c1a5/packages/data/src/components/with-select/index.js#L140)) when we know we're just passing through props directly, to avoid wastefully iterating over the props in the pass-through.

Reply:
I don't like that we have to change the inner component's implementation to adapt to this. I find it ok for higher-order components that need to be very performant but personally prefer to avoid it in regular components.

Reply:
I'd agree in the sense of not surfacing it on the public API, but it seems perfectly fine (and arguably preferable) for internal pass-through components.